srv/salt/_modules/rgw.py

```diff
@@ -40,42 +40,37 @@ def configuration(role):
         role = 'rgw'
     if 'roles' in __pillar__:
         if 'rgw_configurations' in __pillar__:
-            for rgw_config in  __pillar__['rgw_configurations'].keys():
+            for rgw_config in  __pillar__['rgw_configurations']:
                 if rgw_config in role:
                     return rgw_config
-    return 
+    return
 
 
 
-def users(role):
+def users(realm='default'):
     """
-    Return the list of users.  Consider the default rgw and ganesha roles
-    equivalent.
+    Return the list of users for a realm.
     """
-    if 'roles' in __pillar__:
-        if 'rgw_configurations' in __pillar__:
-            if role == 'ganesha' or role == 'rgw':
-                # Special case for default names
-                users = [ u['uid'] for u in __pillar__['rgw_configurations']['rgw']['users'] ]
-                log.info("users: {}".format(users))
-                return users
-            if role in __pillar__['rgw_configurations']:
-                users = [ u['uid'] for u in __pillar__['rgw_configurations'][role]['users'] ]
-                log.info("users: {}".format(users))
-                return users
-        if 'rgw' in __pillar__['roles']:
-            return []
+    cmd = "radosgw-admin user list --rgw-realm={}".format(realm)
+    log.info("cmd: {}".format(cmd))
+    proc = Popen(cmd.split(), stdout=PIPE, stderr=PIPE)
+    proc.wait()
+    log.debug("rc: {}".format(proc.returncode))
+    if proc.returncode != '0':
+        return json.loads(proc.stdout.read())
     return []
 
-def add_users(pathname="/srv/salt/ceph/rgw/cache"):
+def add_users(pathname="/srv/salt/ceph/rgw/cache", jinja="/srv/salt/ceph/rgw/files/users.j2"):
     """
-    Write each user to its own file
+    Write each user to its own file.
     """
-    if 'rgw_configurations' not in __pillar__:
+    users = __salt__['slsutil.renderer'](jinja)
+
+    if users is None or 'realm' not in users:
         return
-    
-    for role in __pillar__['rgw_configurations']:
-        for user in __pillar__['rgw_configurations'][role]['users']:
+
+    for realm in users['realm']:
```

Comment:
From the structure of the file `/srv/salt/ceph/rgw/files/users.j2`, the value of the `realm` key is not a list but a dict, which means that you should change this to:
```
for realm, user_list in users['realm'].items():
    for user in user_list:
```

Reply:
I am hardcoding default for the moment and did not plan to extend this yet.  My ulterior goal is to set us up for rgw multi-site without too much reorganization.

Reply:
I just noticed that iteration over a python dict return the key values, which means that this part of code should work as expected.